Improve performance of grouped iterators

som-snytt opened the pull request:
Cf https://github.com/scala/scala/pull/9818

som-snytt committed:
Avoid Option allocation

som-snytt committed:
Simplify grouped next/hasNext

som-snytt committed:
Simplify grouped buffer fulfillment, clarify docs

som-snytt committed:
Grouped iterator allocs and frees when it can

som-snytt committed:
Move two old partest to junit

joroKr21 commented on src/library/scala/collection/Iterator.scala:
```diff
@@ -155,124 +155,115 @@ trait Iterator[+A] extends IterableOnce[A] with IterableOnceOps[A, Iterator, Ite
+            buffer += pad()
+            index += 1
+          }
+          // TODO by-name pad should be taken as lazy; fix the test
+          //buffer.padToInPlace(size, pad())
+          //index = size
```
Wouldn't that change semantics?

som-snytt replied:
Yes but in a good way. It can't possibly be a good idea to invoke the fill arg N times, perhaps.

joroKr21 replied:
I guess it's very unlikely that someone relies on side effects in padding a grouped iterator 🤷 
I'm not confident to make this decision though.

som-snytt replied:
Maybe it needs a special comment for "we can't do this until the hypothetical future breaking version". There is a comment in TreeSeqMap to similar effect. I will change it from overly forceful TODO to something that expresses WBNTHIF.

joroKr21 commented:
This looks like a really nice improvement - I would delete the comment about `padToInPlace` because I don't think we can just change semantics like that. Otherwise LGTM 👍

som-snytt commented:
It's still too bad for normal grouped case to do extra array copy. Just give me an array.

Also, I would like to pad with zero: just give me the array that is already zeroed out.

joroKr21 commented:
I don't think we can avoid the copy though. 🤔 or am I missing something.
Regarding pad with zero that sounds too edge-casey to me.

som-snytt commented:
For normal grouped (no sliding overlap), it should just allocate an array and fill it, and return it wrapped, rather than laboriously ArrayBuffer which is more abstraction than needed. 

One of the Shipilev blogs about array copying turns on the cost of zeroing the array. But at least here it's only the last segment that is padded. The very word "padding" sounds like a wasteful luxury, like "automatic transmission".

I improved the wording in the commented code; I retained the commented code because it passes the Marie Kondo test.

som-snytt commented:
Reviewing the doc change, leaving a reminder here about what I found confusing: `withPartial` says whether to keep the partial segment, not whether to return a partial segment; whereas I thought:
```
(no partial, no padding) => drop segment
(no partial, padding) => pad segment instead of drop segment
(yes partial, no padding) => keep segment
(yes partial, padding) => pad segment
```
Not sure I tabulated that correctly, but I'm too tired to verify.

Instead of two boolean flags, an alternative API would be `drop`, `keep`, and `pad`. Then also `padNull` for `pad(null.asInstanceOf[A])` to keep default zeroes.

joroKr21 commented:
You are correct 👍

joroKr21 commented on src/library/scala/collection/Iterator.scala:
```diff
@@ -155,124 +155,115 @@ trait Iterator[+A] extends IterableOnce[A] with IterableOnceOps[A, Iterator, Ite
+          buffer += self.next()
+          index += 1
+        }
+        if (partial && pad != null) {
```
I think this is a change of behaviour: the comment before said
>  // If there is padding defined we insert it immediately
    // so the rest of the code can be oblivious

i.e. if there is padding specified, we shouldn't check `partial`. Unless we set `partial = true` in `withPadding`. But also setting `withPartial(true)` explicitly (as opposed to the default) used to remove the padding.

som-snytt replied:
Yes, the change was intentional because the previous behavior was a design bug. I think it was just due to the complexity of the implementation, compare your other comment.

joroKr21 commented on src/library/scala/collection/Iterator.scala:
```diff
@@ -155,124 +155,115 @@ trait Iterator[+A] extends IterableOnce[A] with IterableOnceOps[A, Iterator, Ite
      */
     def withPartial(x: Boolean): this.type = {
-      _partial = x
-      if (_partial) // reset pad since otherwise it will take precedence
-        pad = None
```
I think we should keep this

som-snytt replied:
There's a test to say don't lose the pad function. Since it's "inconceivable" (to quote Princess Bride) that anyone is relying on the old behavior, I made it more sane. (I take the comment to mean, "because of the way I coded it, I have to reset the pad function.") I'm not unwilling to be persuaded either way.

joroKr21 replied:
The way I see it is that the last setting I provided is the one that matters more. If I said I want padding, then partial or not doesn't matter. If I said I want `partial = true` then it doesn't make sense to pad because that would never return a partial result.

som-snytt replied:
Right, I was about to say maybe I inverted the meaning of the settings. Partial means give me my partial segment, otherwise, drop it if no padding, or pad it otherwise.

som-snytt replied:
but that is not order dependent. Partial means no pad.

som-snytt replied:
I was waiting for a benchmark test to run, but it's taking half the night. I want to be able to benchmark before I fiddle with `PriorityQueue` (just tweaking after touching the scaladoc). I will kill the benchmark. I guess I'll run a benchmark test when I leave for vacation and check the results after I return.

som-snytt replied:
Thanks again for your eyes. To recap, my change was to read "withPartial" to mean "keep the partial segment" instead of "return a partial segment".

joroKr21 commented on test/junit/scala/collection/IteratorTest.scala:
```diff
@@ -1,19 +1,173 @@
+    assertTrue(g.hasNext)
+    res = g.next()
+    assertEquals("got a group", 3, res.length)
+    assertFalse(g.hasNext)
```
Currently this pads:
```scala
scala> 1.to(7).iterator.grouped(3).withPartial(false).withPadding(42).toList
val res3: List[Seq[Int]] = List(ArraySeq(1, 2, 3), ArraySeq(4, 5, 6), ArraySeq(7, 42, 42))
```

joroKr21 replied:
Hmm this also pads:
```scala
scala> 1.to(7).iterator.grouped(3).withPadding(42).withPartial(false).toList
val res0: List[Seq[Int]] = List(ArraySeq(1, 2, 3), ArraySeq(4, 5, 6), ArraySeq(7, 42, 42))
```

joroKr21 reviewed (changes requested):
Sorry, just realised that you are slightly changing behaviour wrt to padding vs partial

som-snytt commented:
I updated my previous table of behaviors.

The only difference in behavior should be not to null out the pad function if requesting partial. IIUC.

joroKr21 commented:
> I updated my previous table of behaviors.

What you have currently though is:
(no partial, padding) => drop segment // edit; it used to pad

This is another change apart from nulling out the padding.

joroKr21 reviewed (commented):
> Yes, the change was intentional because the previous behavior was a design bug. I think it was just due to the complexity of the implementation, compare your other comment.

Ok, I think that's not so bad - I just didn't notice the change originally but I'm actually happy with removing the order-dependence between `withPartial` and `withPadding`.

som-snytt commented:
Because of the new Matrix movie, I'm re-watching Point Break (1991). I didn't like Keanu's turn as Don John in Much Ado About Nothing, and he's also wooden here when he pretends like "I have a plan", but Patrick Swayze is really that good and Keanu is better when he loses the tie, much like in Matrix.

Sorry to contradict you, but you were right: If you ask for padding, you get padding if necessary; otherwise, if you don't want partial segment, drop it; otherwise just return it. My functional change is just not to null out the pad function, which is an edge which never happens.

I'll push that when Point Break goes to commercial break. Apparently, the pad function is not the Option, rather the partial flag is the Option! because we need to differentiate "user asked for partial".

som-snytt committed:
GroupedIterator withPartial and withPadding are exclusive

som-snytt committed:
Simplify grouped tests

som-snytt committed:
Improve stack depth test of ConcatIterator

som-snytt commented:
To summarize, integrating @joroKr21 observations, the changed behavior is now that padding is canceled on any invocation of `withPartial`. The consistency makes destructive `withPartial(true)` less mysterious. The doc is more true, that the two methods are "mutually exclusive". Probably no one in the history of Scala has abused this API anyway. Famous last words, but how else ought one speak as we approach the Apocalypse?
```
➜  ~ scala
Welcome to Scala 2.13.7 (OpenJDK 64-Bit Server VM, Java 17).
Type in expressions for evaluation. Or try :help.

scala> (1 to 7).iterator.grouped(3).withPadding(42).withPartial(false).toList
val res0: List[Seq[Int]] = List(ArraySeq(1, 2, 3), ArraySeq(4, 5, 6), ArraySeq(7, 42, 42))

scala> (1 to 7).iterator.grouped(3).withPadding(42).withPartial(true).withPartial(false).toList
val res1: List[Seq[Int]] = List(ArraySeq(1, 2, 3), ArraySeq(4, 5, 6))

scala>
:quit
➜  ~ skala
Welcome to Scala 2.13.8-20211221-204044-00c1895 (OpenJDK 64-Bit Server VM, Java 17).
Type in expressions for evaluation. Or try :help.

scala> (1 to 7).iterator.grouped(3).withPadding(42).withPartial(false).toList
val res0: List[Seq[Int]] = List(ArraySeq(1, 2, 3), ArraySeq(4, 5, 6))

scala> (1 to 7).iterator.grouped(3).withPadding(42).withPartial(true).withPartial(false).toList
val res1: List[Seq[Int]] = List(ArraySeq(1, 2, 3), ArraySeq(4, 5, 6))

scala>
```

som-snytt committed:
GroupedIterator uses ArrayBuilder

Previous use of ArrayBuffer would shift the overlap from stepping,
fill the remainder, then toArray for result, then either retain
the buffer if stepping or clear it.

Now it allocates the result, copies the overlap, fills and returns it.
If stepping, an array of just overlap size is allocated and retained.
The retained buffer is never cleared, but reference is nulled when done.

som-snytt commented:
Simple benchmark before/after says using array builder vs buffer did something.
```
[info] Benchmark                  (size)  Mode  Cnt   Score   Error  Units
[info] IteratorBenchmark.grouped   10000  avgt   30  12.714 ± 0.078  ns/op
[info] IteratorBenchmark.sliding   10000  avgt   30  12.364 ± 0.265  ns/op

[info] IteratorBenchmark.grouped   10000  avgt   30  5.373 ± 0.050  ns/op
[info] IteratorBenchmark.sliding   10000  avgt   30  4.951 ± 0.245  ns/op
```
When I understand benchmarking better, I'll re-evaluate.

joroKr21 reviewed (approved):
Very smart 👍

SethTisue commented:
Let's wait a few days to see if @scala/collections crew has any feedback

Ichoran commented:
I had a quick look through and it seems good to me, but I can't vouch that I looked carefully enough to catch subtle problems.

som-snytt commented:
and all the problems are subtle.

If you break `sliding`, you can no longer bootstrap the compiler! True story.

I did a few rounds of benchmarking but not expert yet. The difference was in the correct direction and was significant as in factor of 2, i.e., significant enough to make me press on.

I'll PR the benchmarks at some point, nothing special.

som-snytt commented:
> Let's wait a few days to see if @scala/collections crew has any feedback

That was encouraging feedback from joroKollections.

som-snytt commented:
Thanks @joroKr21 for starting this initiative and also supporting with further eyeballs. The result was much improved, although any remaining deficiencies are my own, as people say when thanking their editors and early reviewers.

sjrd commented:
It seems this PR broke the test for `js.ArrayOps.sliding`:
https://github.com/scala-js/scala-js/blob/c27013fac5089ce476efe934fb55e6d525a912e4/test-suite/js/src/test/scala/org/scalajs/testsuite/library/ArrayOpsTest.scala#L754-L760
The returned arrays are now:
```
js.Array(1, 2, 1, 3)
js.Array(3, 1, 10, 9, 3)
js.Array(0)
```
instead of the expected
```
js.Array(1, 2, 1, 3)
js.Array(3, 1, 10, 9)
js.Array(9, 3, 0)
```
It seems the overlapping aspect is lost.

`js.Array.sliding` is provided as an implicit class here:
https://github.com/scala-js/scala-js/blob/c27013fac5089ce476efe934fb55e6d525a912e4/library/src/main/scala-new-collections/scala/scalajs/js/ArrayOps.scala#L1528-L1529
and that refers to `js.WrappedArray`, which is here:
https://github.com/scala-js/scala-js/blob/c27013fac5089ce476efe934fb55e6d525a912e4/library/src/main/scala-new-collections/scala/scalajs/js/WrappedArray.scala#L24

I have not yet managed to determine whether the new implementation in this PR is at fault, or whether we were already doing something wrong, that is now surfaced by this new implementation.

Does the above output ring any bell to the people involved in this PR?

joroKr21 commented:
Maybe some difference in the behaviour of `Array.copy`?
```scala
        if (step < size) {
          if (first) prev = buffer.drop(step)
          else if (buffer.length == size) Array.copy(src = buffer, srcPos = step, dest = prev, destPos = 0, length = size - step)
          else prev = null
        }
```

sjrd commented:
That does not seem likely. It's literally the same code that is used for `scala.Array.copy` in JS and on the JVM. Besides, it is very unlikely that it can explain why the *lengths* of the arrays are different.

sjrd commented:
So, it is our fault. The problem is that we have [our own implementation of `ArrayBuilder`](https://github.com/scala-js/scala-js/blob/11e35c16c1740031d34ea5b8d595cfefde7c32c5/scalalib/overrides-2.13/scala/collection/mutable/ArrayBuilder.scala#L94-L148), but its `length` method was wrong. We did not notice, when adding support for 2.13, that new members had been added to `ArrayBuilder`, and so we forgot to override `def length = size`. We don't use the `size` field; we directly use the underlying `jsArray.length`.

joroKr21 commented:
Wow, nice finding 👍 - does it make sense to use `.size` here to mitigate the issue?

sjrd commented:
No, that would be worse, as we can't fix our `size` to evolve. We need to override `length` and `knownSize` to use something else than `size` (see https://github.com/scala-js/scala-js/pull/4628). Also, `size` is a `protected var`, so you can't use it anyway.

joroKr21 commented:
In principle we do know the size externally but we would have to expend an extra field to track it or otherwise significantly restructure the code.

som-snytt commented:
OK I'm awake. Thanks for diagnosing and fixing during my overnight. I'm sorry to have missed out, as I recently got my feet wet again with scala-js by wading in on an issue.

This failure makes me wonder about component selection and complexity: swapping otherwise-equivalent `A` for `B` incurs at least a testing burden.

(Also more to consider cross-platform; there is an unrelated ticket about performance on scala-native.)

Looking back, I think the PR is correct to call it a simplification, although it's hard to tell from the diff.
